core/shared/src/main/scala/zio/stm/TSemaphore.scala

```diff
@@ -43,14 +91,54 @@ final class TSemaphore private (val permits: TRef[Long]) extends AnyVal {
       permits.unsafeSet(journal, current + n)
     }
 
-  def withPermit[E, B](stm: STM[E, B]): STM[E, B] =
-    acquire *> stm <* release
+  /**
+   * Executes the specified effect, acquiring a permit immediately before the
+   * effect begins execution and releasing it immediately after the effect
+   * completes execution, whether by success, failure, or interruption.
+   */
+  def withPermit[R, E, A](zio: ZIO[R, E, A]): ZIO[R, E, A] =
+    withPermits(1L)(zio)
```

Comment:
Does it make sesne to keep the old `withPermit`?

Reply:
One of the challenges of this is creating an API that feels natural both when you are using it as just a `Semaphore` and when you are using it as a `TSemaphore`.

I thought making the `withPermit` variants take a `ZIO` effect and having everything else return an `STM` effect struck a nice balance because `withPermit` variants were the only ones used with `Semaphore` before and created a clear communication message of `withPermit` for basic stuff, other operators for fancy stuff.

Also I think for STM there is less need for it because you don't have to worry about interruption within STM and I think most of the time if you are using STM you want to compose it with other things.

So I would be inclined to keep it as is but maybe there is a use case I am missing.

Reply:
Agreed, let's keep it this way.